File: src/talker_listener_python/talker_listener_python/listener.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Listener(Node):
    def __init__(self):
        super().__init__("listener")
        self.subscriber = self.create_subscription(
            String, "topic", self.listener_callback, 10
        )
        self.subscriber  # prevent unused warning.

    def listener_callback(self, msg: String):
        self.get_logger().info(f"received {msg}")


def main(args=None):
    rclpy.init(args=args)
    listener = Listener()
    rclpy.spin(listener)

File: src/talker_listener_cpp/src/listener.cpp
#include <cstdio>
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"  // "" means search local first, mostly used for self-defined files..
#include "std_msgs/msg/string.hpp"

using std::placeholders::_1;

class Listener : public rclcpp::Node
{
public:
  Listener() : Node("Listener")
  {
    subscription_ =
        this->create_subscription<std_msgs::msg::String>("topic", 10, std::bind(&Listener::topic_callback, this, _1));
  }

private:
  void topic_callback(const std_msgs::msg::String::SharedPtr msg)
  {
    RCLCPP_INFO(this->get_logger(), "received '%s' ", msg->data.c_str());
  }
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
  return 0;
}
